contracts/utils/TokenVestingStaking.sol

```diff
@@ -146,6 +168,51 @@ contract TokenVestingStaking is ITokenVestingStaking, Shared {
         SENT.safeTransfer(beneficiary, amount);
     }
 
+        /**
+     * @notice Contributes funds to a ServiceNodeContribution contract
+     * @param contributionContract The address of the ServiceNodeContribution contract
+     * @param amount The amount of SENT tokens to contribute
+     */
+    function contributeFunds(address contributionContract, uint256 amount) external onlyBeneficiary notRevoked afterStart {
+        bool contractDeployed = contributionFactory.isContractDeployed(contributionContract);
+        require(contractDeployed, "Invalid contribution contract");
+        IServiceNodeContribution contribution = IServiceNodeContribution(contributionContract);
+
+        bool isNewNode = true;
+        for (uint256 i = 0; i < investorServiceNodes.length; i++) {
+            if (investorServiceNodes[i].contributionContract == contributionContract) {
+                investorServiceNodes[i].deposit += amount;
+                isNewNode = false;
+                break;
+            }
+        }
+        if (isNewNode)
+            investorServiceNodes.push(ServiceNode(0, amount, contributionContract));
+
+        SENT.approve(contributionContract, amount);
+        contribution.contributeFunds(amount);
+    }
+
+    /**
+     * @notice Withdraws contribution from a ServiceNodeContribution contract
+     * @param contributionContractAddress The address of the ServiceNodeContribution contract
+     */
+    function withdrawContribution(address contributionContractAddress) external onlyBeneficiary notRevoked afterStart {
+        require(contributionFactory.isContractDeployed(contributionContractAddress), "Invalid contribution contract");
+        IServiceNodeContribution contributionContract = IServiceNodeContribution(contributionContractAddress);
+        bool foundNode = false;
+        for (uint256 i = 1; i < investorServiceNodes.length + 1; i++) {
+            if (investorServiceNodes[i - 1].contributionContract == contributionContractAddress) {
```

Comment:
Similar stale node problem here if the contract is reset/cancelled or reused. The entry will persist and there won't be a way to get rid of it. I think we can just flush them out on claimRewards, or all these functions that interact with the contract have a modifier to groom the list.